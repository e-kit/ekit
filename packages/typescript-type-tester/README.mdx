---
name: TypeScript Type Tester
menu: '开发/测试/构建'
route: /ekit/typescript-type-tester
---

# TypeScript Type Tester

TypeScript 类型测试工具。

```
npm i -S @ekit/typescript-type-teste
```

## parser

通过检测示例代码内的 TypeScript 类型错误来进行对自定义工具类型或者 TypeScript 代码类型的单元测试，示例：

### xxOK.ts

```ts
// 符合静态类型检测，不应该抛出错误
const str: string = 'string';
const num: number = 1;
```

### xxFail.ts

```ts
// 不符合静态类型检测，应该抛出错误
const str: string = 1;
const num: number = 'string';
```

### xx.spec.ts 单测代码

```ts
import { parser } from '@ekit/typescript-type-tester';

describe('xx work ok', () => {
  const files = 'xxFail.ts xxOK.ts'.split(' ');
  const errors = parser(files, '__tests__');
  files.forEach((file) =>
    it(`check ${file} work ok`, () => {
      const current = errors[file];
      if (file.match(/OK/)) {
        expect(current).toHaveLength(0);
      } else if (file.match(/Fail/)) {
        expect(current.length).toBeGreaterThan(0);
      }
      expect(current).toMatchSnapshot();
    })
  );
});
```
