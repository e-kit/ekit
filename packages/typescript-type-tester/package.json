{
  "name": "@ekit/typescript-type-tester",
  "version": "0.0.4",
  "description": "test typescript type with jest and typescript",
  "keywords": [
    "typescript",
    "jest"
  ],
  "author": "yangqianjun",
  "homepage": "https://github.com/e-kit/ekittree/master/packages/typescript-type-tester#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gogoyqj/ekit.git"
  },
  "scripts": {
    "test": "jest --all",
    "build": "tsc --project tsconfig.prod.json"
  },
  "bugs": {
    "url": "https://github.com/e-kit/ekitissues"
  },
  "peerDependencies": {
    "tslib": "*",
    "typescript": "^3.8.2"
  },
  "gitHead": "b395eb7c33153865c8fe74ff270b814d08805308"
}
