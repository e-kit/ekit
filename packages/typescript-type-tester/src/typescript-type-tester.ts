import * as path from 'path';
import * as ts from 'typescript';

/**
 * 解析一个 .ts 文件并返回错误信息
 * @param files 文件名
 * @param directory 绝对路径
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parser = (
  files: string[],
  directory: string,
  runtimeCompilerOptions: ts.CompilerOptions = {}
) => {
  /** 当前执行脚本的目录 */
  const cwd = process.cwd();
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { compilerOptions } = require(path.join(cwd, 'tsconfig.json'));
  const fileNameMap: { [file: string]: any[] } = {};
  const errorMap: { [file: string]: any[] } = {};
  const program = ts.createProgram(
    files.map((file) => {
      const p = path.join(directory, file);
      fileNameMap[p] = errorMap[file] = [];
      return p;
    }),
    {
      ...compilerOptions,
      ...runtimeCompilerOptions,
      moduleResolution: ts.ModuleResolutionKind.NodeJs,
    }
  );
  const allDiagnostics = ts.getPreEmitDiagnostics(program);
  allDiagnostics.map((d) => {
    if (d.file) {
      if (d.file.fileName in fileNameMap) {
        const errors = fileNameMap[d.file.fileName];
        const { relatedInformation, code, category, messageText } = d;
        errors.push({
          fileName: path.relative(cwd, d.file.fileName),
          code,
          category,
          messageText: ts.flattenDiagnosticMessageText(messageText, '\n'),
        });
        if (Array.isArray(relatedInformation)) {
          relatedInformation.forEach((d) => {
            const { code, category, messageText, file } = d;
            // 方便测试
            console.log({
              fileName: path.relative(cwd, (file && file.fileName) || ''),
              code,
              category,
              messageText: ts.flattenDiagnosticMessageText(messageText, '\n'),
            });
          });
        }
      }
    }
  });
  return errorMap;
};
