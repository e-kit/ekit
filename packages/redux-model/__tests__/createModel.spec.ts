/**
 * @file: createModel 语法测试
 */
import { parser } from '@ekit/typescript-type-tester';

describe('utils/createModel work ok', () => {
  const files = 'testCreateModelFail.ts testCreateModelOK.ts testCreateModelStateReadOnlyFail.ts testCreateModelStateReadOnlyOK.ts testCreateModelTYPESFail.ts testCreateModelTYPESOK.ts testPutCallInModelFail.ts testPutCallInModelOK.ts testPutCallOutModelFail.ts testPutCallOutModelOK.ts testPutCallWithOverloadFail.ts testPutCallWithOverloadOK.ts testWithEffectsPutAndCallFail.ts testWithEffectsPutAndCallOK.ts'.split(
    ' '
  );
  const errors = parser(files, '__tests__/createModelSamples');
  files.forEach((file) =>
    it(`check sample/${file} work ok`, () => {
      const current = errors[file];
      if (file.match(/OK/)) {
        expect(current).toHaveLength(0);
      } else if (file.match(/Fail/)) {
        expect(current.length).toBeGreaterThan(0);
      }
      expect(current).toMatchSnapshot();
    })
  );
});

describe('utils/createModel async work ok', () => {
  const files = 'testCreateAsyncModelFail.ts testCreateAsyncModelOK.ts testCreateMixAsyncModelFail.ts testCreateMixAsyncModelOK.ts'.split(
    ' '
  );
  const errors = parser(files, '__tests__/asyncFactorySamples');
  files.forEach((file) =>
    it(`check sample/${file} work ok`, () => {
      const current = errors[file];
      if (file.match(/OK/)) {
        expect(current).toHaveLength(0);
      } else if (file.match(/Fail/)) {
        expect(current.length).toBeGreaterThan(0);
      }
      expect(current).toMatchSnapshot();
    })
  );
});
