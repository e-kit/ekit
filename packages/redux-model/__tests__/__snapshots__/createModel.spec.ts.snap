// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/createModel async work ok check sample/testCreateAsyncModelFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/asyncFactorySamples/testCreateAsyncModelFail.ts",
    "messageText": "Type 'number' is not assignable to type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/asyncFactorySamples/testCreateAsyncModelFail.ts",
    "messageText": "Type 'string' is not assignable to type 'number'.",
  },
]
`;

exports[
  `utils/createModel async work ok check sample/testCreateAsyncModelOK.ts work ok 1`
] = `Array []`;

exports[`utils/createModel async work ok check sample/testCreateMixAsyncModelFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/asyncFactorySamples/testCreateMixAsyncModelFail.ts",
    "messageText": "Type 'number' is not assignable to type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/asyncFactorySamples/testCreateMixAsyncModelFail.ts",
    "messageText": "Type 'string' is not assignable to type 'number'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/asyncFactorySamples/testCreateMixAsyncModelFail.ts",
    "messageText": "Type 'string' is not assignable to type 'number'.",
  },
]
`;

exports[
  `utils/createModel async work ok check sample/testCreateMixAsyncModelOK.ts work ok 1`
] = `Array []`;

exports[`utils/createModel work ok check sample/testCreateModelFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/createModelSamples/testCreateModelFail.ts",
    "messageText": "Type 'string' is not assignable to type 'number'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/createModelSamples/testCreateModelFail.ts",
    "messageText": "Type '<P extends AbstractAction>(state: Readonly<State>, action: Tction<string>) => { id: string; name: string; }' is not assignable to type '<P extends AbstractAction>(state: Readonly<State>, action: P) => State'.
  Call signature return types '{ id: string; name: string; }' and 'State' are incompatible.
    The types of 'id' are incompatible between these types.
      Type 'string' is not assignable to type 'number'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/createModelSamples/testCreateModelFail.ts",
    "messageText": "Type '<P extends AbstractAction>(state: Readonly<State>, action: Tction<{ user: { id: string; }; }>) => { id: string; name: string; }' is not assignable to type '<P extends AbstractAction>(state: Readonly<State>, action: P) => State'.
  Call signature return types '{ id: string; name: string; }' and 'State' are incompatible.
    The types of 'id' are incompatible between these types.
      Type 'string' is not assignable to type 'number'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/createModelSamples/testCreateModelFail.ts",
    "messageText": "Type 'number' is not assignable to type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/createModelSamples/testCreateModelFail.ts",
    "messageText": "Type '<P extends AbstractAction>(state: Readonly<State>, action: Tction<number>) => { name: number; id: number; }' is not assignable to type '<P extends AbstractAction>(state: Readonly<State>, action: P) => State'.
  Call signature return types '{ name: number; id: number; }' and 'State' are incompatible.
    The types of 'name' are incompatible between these types.
      Type 'number' is not assignable to type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/createModelSamples/testCreateModelFail.ts",
    "messageText": "Type '<P extends AbstractAction>(state: Readonly<State>, action: Tction<{ user: { name: number; }; }>) => { name: number; id: number; }' is not assignable to type '<P extends AbstractAction>(state: Readonly<State>, action: P) => State'.
  Call signature return types '{ name: number; id: number; }' and 'State' are incompatible.
    The types of 'name' are incompatible between these types.
      Type 'number' is not assignable to type 'string'.",
  },
]
`;

exports[`utils/createModel work ok check sample/testCreateModelOK.ts work ok 1`] = `Array []`;

exports[`utils/createModel work ok check sample/testCreateModelStateReadOnlyFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2540,
    "fileName": "__tests__/createModelSamples/testCreateModelStateReadOnlyFail.ts",
    "messageText": "Cannot assign to 'id' because it is a read-only property.",
  },
]
`;

exports[
  `utils/createModel work ok check sample/testCreateModelStateReadOnlyOK.ts work ok 1`
] = `Array []`;

exports[`utils/createModel work ok check sample/testCreateModelTYPESFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2339,
    "fileName": "__tests__/createModelSamples/testCreateModelTYPESFail.ts",
    "messageText": "Property 'a' does not exist on type '{ testNoArguments: string; testOneArguments: string; } & { testAsyncNoArguments: string; testAsyncOneArguments: string; }'.",
  },
]
`;

exports[`utils/createModel work ok check sample/testCreateModelTYPESOK.ts work ok 1`] = `Array []`;

exports[`utils/createModel work ok check sample/testPutCallInModelFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallInModelFail.ts",
    "messageText": "Argument of type '\\"\\"' is not assignable to parameter of type 'never'.",
  },
  Object {
    "category": 1,
    "code": 2769,
    "fileName": "__tests__/createModelSamples/testPutCallInModelFail.ts",
    "messageText": "No overload matches this call.
  Overload 1 of 4, '(effect: string | Tction<any>): Promise<unknown>', gave the following error.
    Argument of type '<P_1 extends Tction<string>>(payload: P_1[\\"payload\\"]) => Promise<number>' is not assignable to parameter of type 'string | Tction<any>'.
      Property 'payload' is missing in type '<P_1 extends Tction<string>>(payload: P_1[\\"payload\\"]) => Promise<number>' but required in type 'Tction<any>'.
  Overload 2 of 4, '(effect: () => any): Promise<any>', gave the following error.
    Argument of type '<P_1 extends Tction<string>>(payload: P_1[\\"payload\\"]) => Promise<number>' is not assignable to parameter of type '() => any'.",
  },
]
`;

exports[`utils/createModel work ok check sample/testPutCallInModelOK.ts work ok 1`] = `Array []`;

exports[`utils/createModel work ok check sample/testPutCallOutModelFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallOutModelFail.ts",
    "messageText": "Argument of type '\\"\\"' is not assignable to parameter of type 'never'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallOutModelFail.ts",
    "messageText": "Argument of type '\\"\\"' is not assignable to parameter of type 'never'.",
  },
  Object {
    "category": 1,
    "code": 2769,
    "fileName": "__tests__/createModelSamples/testPutCallOutModelFail.ts",
    "messageText": "No overload matches this call.
  Overload 1 of 4, '(effect: string | Tction<any>): Promise<unknown>', gave the following error.
    Argument of type '<P extends Tction<string>>(payload: P[\\"payload\\"]) => P' is not assignable to parameter of type 'string | Tction<any>'.
      Property 'payload' is missing in type '<P extends Tction<string>>(payload: P[\\"payload\\"]) => P' but required in type 'Tction<any>'.
  Overload 2 of 4, '(effect: () => any): Promise<any>', gave the following error.
    Argument of type '<P extends Tction<string>>(payload: P[\\"payload\\"]) => P' is not assignable to parameter of type '() => any'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallOutModelFail.ts",
    "messageText": "Argument of type '<P extends Tction<string>>(payload: P[\\"payload\\"]) => P' is not assignable to parameter of type '() => any'.",
  },
]
`;

exports[`utils/createModel work ok check sample/testPutCallOutModelOK.ts work ok 1`] = `Array []`;

exports[`utils/createModel work ok check sample/testPutCallWithOverloadFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallWithOverloadFail.ts",
    "messageText": "Argument of type '2' is not assignable to parameter of type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallWithOverloadFail.ts",
    "messageText": "Argument of type '2' is not assignable to parameter of type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallWithOverloadFail.ts",
    "messageText": "Argument of type '\\"\\"' is not assignable to parameter of type 'number'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testPutCallWithOverloadFail.ts",
    "messageText": "Argument of type '2' is not assignable to parameter of type 'string'.",
  },
]
`;

exports[
  `utils/createModel work ok check sample/testPutCallWithOverloadOK.ts work ok 1`
] = `Array []`;

exports[`utils/createModel work ok check sample/testWithEffectsPutAndCallFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testWithEffectsPutAndCallFail.ts",
    "messageText": "Argument of type '\\"\\"' is not assignable to parameter of type 'never'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/createModelSamples/testWithEffectsPutAndCallFail.ts",
    "messageText": "Argument of type '\\"\\"' is not assignable to parameter of type 'never'.",
  },
  Object {
    "category": 1,
    "code": 2554,
    "fileName": "__tests__/createModelSamples/testWithEffectsPutAndCallFail.ts",
    "messageText": "Expected 2 arguments, but got 3.",
  },
  Object {
    "category": 1,
    "code": 2554,
    "fileName": "__tests__/createModelSamples/testWithEffectsPutAndCallFail.ts",
    "messageText": "Expected 2 arguments, but got 3.",
  },
]
`;

exports[
  `utils/createModel work ok check sample/testWithEffectsPutAndCallOK.ts work ok 1`
] = `Array []`;
